File: packages/memsh-cli/src/tool/edit.ts
import { z } from "zod"
import { Tool } from "./tool"

const DESCRIPTION = `Performs exact string replacements in files in the memsh in-memory filesystem.

Usage:
- The edit will FAIL if oldString is not found in the file
- The edit will FAIL if oldString is not unique in the file (unless replaceAll is true)
- Use replaceAll for replacing and renaming strings across the file
- If oldString is empty, the file will be created with newString as content`

export interface EditMetadata {
  filepath: string
  diff: string
  additions: number
  deletions: number
  [key: string]: unknown
}

/**
 * Create a simple unified diff
 */
function createDiff(filepath: string, oldContent: string, newContent: string): string {
  const oldLines = oldContent.split("\n")
  const newLines = newContent.split("\n")

  let diff = `--- ${filepath}\n+++ ${filepath}\n`

  // Simple diff: show removed and added lines
  const maxLines = Math.max(oldLines.length, newLines.length)
  let inHunk = false
  let hunkStart = 0
  let hunkLines: string[] = []

  const flushHunk = () => {
    if (hunkLines.length > 0) {
      diff += `@@ -${hunkStart + 1} +${hunkStart + 1} @@\n`
      diff += hunkLines.join("\n") + "\n"
      hunkLines = []
    }
    inHunk = false
  }

  for (let i = 0; i < maxLines; i++) {
    const oldLine = oldLines[i]
    const newLine = newLines[i]

    if (oldLine === newLine) {
      if (inHunk) {
        hunkLines.push(` ${oldLine ?? ""}`)
        if (hunkLines.filter((l) => l.startsWith(" ")).length > 3) {
          flushHunk()
        }
      }
    } else {
      if (!inHunk) {
        inHunk = true
        hunkStart = Math.max(0, i - 3)
        // Add context before
        for (let j = hunkStart; j < i; j++) {
          if (oldLines[j] !== undefined) {
            hunkLines.push(` ${oldLines[j]}`)
          }
        }
      }
      if (oldLine !== undefined && (newLine === undefined || oldLine !== newLine)) {
        hunkLines.push(`-${oldLine}`)
      }
      if (newLine !== undefined && (oldLine === undefined || oldLine !== newLine)) {
        hunkLines.push(`+${newLine}`)
      }
    }
  }

  flushHunk()

  return diff
}

/**
 * Count additions and deletions
 */
function countChanges(oldContent: string, newContent: string): { additions: number; deletions: number } {
  const oldLines = oldContent.split("\n")
  const newLines = newContent.split("\n")

  let additions = 0
  let deletions = 0

  // Simple counting: lines that differ
  const maxLines = Math.max(oldLines.length, newLines.length)
  for (let i = 0; i < maxLines; i++) {
    if (oldLines[i] !== newLines[i]) {
      if (oldLines[i] !== undefined) deletions++
      if (newLines[i] !== undefined) additions++
    }
  }

  return { additions, deletions }
}

export const EditTool = Tool.define<
  z.ZodObject<{
    filePath: z.ZodString
    oldString: z.ZodString
    newString: z.ZodString
    replaceAll: z.ZodOptional<z.ZodBoolean>
  }>,
  EditMetadata
>("edit", {
  description: DESCRIPTION,
  parameters: z.object({
    filePath: z.string().describe("The path to the file to modify"),
    oldString: z.string().describe("The text to replace"),
    newString: z.string().describe("The text to replace it with (must be different from oldString)"),
    replaceAll: z.boolean().optional().describe("Replace all occurrences of oldString (default false)"),
  }),
  async execute(params, ctx) {
    const filepath = params.filePath

    if (params.oldString === params.newString) {
      throw new Error("oldString and newString must be different")
    }

    // Handle creating new file when oldString is empty
    if (params.oldString === "") {
      // Create or overwrite file
      const parentDir = filepath.split("/").slice(0, -1).join("/")
      if (parentDir) {
        const parentExists = await ctx.session.exists(parentDir)
        if (!parentExists) {
          await ctx.session.mkdir(parentDir, { recursive: true })
        }
      }

      await ctx.session.writeFile(filepath, params.newString)

      const diff = createDiff(filepath, "", params.newString)
      const { additions, deletions } = countChanges("", params.newString)

      return {
        title: filepath,
        metadata: {
          filepath,
          diff,
          additions,
          deletions,
        },
        output: diff,
      }
    }

    // Check if file exists
    const exists = await ctx.session.exists(filepath)
    if (!exists) {
      throw new Error(`File not found: ${filepath}`)
    }

    // Check if it's a directory
    const isDir = await ctx.session.isDirectory(filepath)
    if (isDir) {
      throw new Error(`Path is a directory, not a file: ${filepath}`)
    }

    // Read current content
    const oldContent = await ctx.session.readFile(filepath)

    // Check if oldString exists in content
    if (!oldContent.includes(params.oldString)) {
      throw new Error("oldString not found in content")
    }

    // Check for multiple occurrences if not replaceAll
    if (!params.replaceAll) {
      const firstIndex = oldContent.indexOf(params.oldString)
      const lastIndex = oldContent.lastIndexOf(params.oldString)

      if (firstIndex !== lastIndex) {
        throw new Error(
          "Found multiple matches for oldString. Provide more surrounding lines in oldString to identify the correct match, or use replaceAll to replace all occurrences.",
        )
      }
    }

    // Perform replacement
    const newContent = params.replaceAll
      ? oldContent.replaceAll(params.oldString, params.newString)
      : oldContent.replace(params.oldString, params.newString)

    // Write updated content
    await ctx.session.writeFile(filepath, newContent)

    // Generate diff
    const diff = createDiff(filepath, oldContent, newContent)
    const { additions, deletions } = countChanges(oldContent, newContent)

    return {
      title: filepath,
      metadata: {
        filepath,
        diff,
        additions,
        deletions,
      },
      output: diff,
    }
  },
})

File: go-opencode/internal/tool/edit.go
package tool

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/agnivade/levenshtein"
	einotool "github.com/cloudwego/eino/components/tool"
	"github.com/opencode-ai/opencode/internal/event"
)

const editDescription = `Performs exact string replacements in files.

Usage:
- The file_path parameter must be an absolute path
- The old_string must exist in the file (exact match required)
- The new_string will replace old_string
- Use replace_all to replace all occurrences
- The edit will FAIL if old_string is not unique (unless using replace_all)`

// EditTool implements file editing.
type EditTool struct {
	workDir string
}

// EditInput represents the input for the edit tool.
type EditInput struct {
	FilePath   string `json:"filePath"`
	OldString  string `json:"oldString"`
	NewString  string `json:"newString"`
	ReplaceAll bool   `json:"replaceAll,omitempty"`
}

// NewEditTool creates a new edit tool.
func NewEditTool(workDir string) *EditTool {
	return &EditTool{workDir: workDir}
}

func (t *EditTool) ID() string          { return "edit" }
func (t *EditTool) Description() string { return editDescription }

func (t *EditTool) Parameters() json.RawMessage {
	return json.RawMessage(`{
		"type": "object",
		"properties": {
			"filePath": {
				"type": "string",
				"description": "The absolute path to the file to edit"
			},
			"oldString": {
				"type": "string",
				"description": "The exact text to replace"
			},
			"newString": {
				"type": "string",
				"description": "The text to replace it with"
			},
			"replaceAll": {
				"type": "boolean",
				"description": "Replace all occurrences (default: false)"
			}
		},
		"required": ["filePath", "oldString", "newString"]
	}`)
}

func (t *EditTool) Execute(ctx context.Context, input json.RawMessage, toolCtx *Context) (*Result, error) {
	var params EditInput
	if err := json.Unmarshal(input, &params); err != nil {
		return nil, fmt.Errorf("invalid input: %w", err)
	}

	if params.OldString == params.NewString {
		return nil, fmt.Errorf("old_string and new_string must be different")
	}

	// Read file
	content, err := os.ReadFile(params.FilePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read file: %w", err)
	}

	text := string(content)

	// Try exact match first
	var newText string
	var count int

	if params.ReplaceAll {
		count = strings.Count(text, params.OldString)
		if count == 0 {
			return t.fuzzyReplace(text, params, toolCtx)
		}
		newText = strings.ReplaceAll(text, params.OldString, params.NewString)
	} else {
		count = strings.Count(text, params.OldString)
		if count == 0 {
			return t.fuzzyReplace(text, params, toolCtx)
		}
		if count > 1 {
			return nil, fmt.Errorf("old_string appears %d times in file. Use replace_all or provide more context", count)
		}
		newText = strings.Replace(text, params.OldString, params.NewString, 1)
		count = 1
	}

	// Write file
	if err := os.WriteFile(params.FilePath, []byte(newText), 0644); err != nil {
		return nil, fmt.Errorf("failed to write file: %w", err)
	}

	// Publish event (SDK compatible: just file path)
	if toolCtx != nil && toolCtx.SessionID != "" {
		event.Publish(event.Event{
			Type: event.FileEdited,
			Data: event.FileEditedData{
				File: params.FilePath,
			},
		})
	}

	return &Result{
		Title:  fmt.Sprintf("Edited %s", filepath.Base(params.FilePath)),
		Output: fmt.Sprintf("Replaced %d occurrence(s)", count),
		Metadata: map[string]any{
			"file":         params.FilePath,
			"replacements": count,
		},
	}, nil
}

// fuzzyReplace attempts to find similar text when exact match fails.
func (t *EditTool) fuzzyReplace(text string, params EditInput, toolCtx *Context) (*Result, error) {
	// Try line-normalized matching
	normalizedOld := normalizeLineEndings(params.OldString)
	normalizedText := normalizeLineEndings(text)

	if strings.Contains(normalizedText, normalizedOld) {
		newText := strings.Replace(normalizedText, normalizedOld, params.NewString, 1)
		if err := os.WriteFile(params.FilePath, []byte(newText), 0644); err != nil {
			return nil, fmt.Errorf("failed to write file: %w", err)
		}

		// Publish event (SDK compatible: just file path)
		if toolCtx != nil && toolCtx.SessionID != "" {
			event.Publish(event.Event{
				Type: event.FileEdited,
				Data: event.FileEditedData{
					File: params.FilePath,
				},
			})
		}

		return &Result{
			Title:  fmt.Sprintf("Edited %s (normalized)", filepath.Base(params.FilePath)),
			Output: "Replaced 1 occurrence (with line ending normalization)",
		}, nil
	}

	// Try fuzzy matching with similarity
	match, similarity := findBestMatch(text, params.OldString)
	if match != "" && similarity >= 0.7 {
		newText := strings.Replace(text, match, params.NewString, 1)
		if err := os.WriteFile(params.FilePath, []byte(newText), 0644); err != nil {
			return nil, fmt.Errorf("failed to write file: %w", err)
		}

		// Publish event (SDK compatible: just file path)
		if toolCtx != nil && toolCtx.SessionID != "" {
			event.Publish(event.Event{
				Type: event.FileEdited,
				Data: event.FileEditedData{
					File: params.FilePath,
				},
			})
		}

		return &Result{
			Title:  fmt.Sprintf("Edited %s (fuzzy)", filepath.Base(params.FilePath)),
			Output: fmt.Sprintf("Replaced 1 occurrence (%.0f%% similarity)", similarity*100),
		}, nil
	}

	return nil, fmt.Errorf("old_string not found in file. The content may have changed or the string doesn't exist")
}

func normalizeLineEndings(s string) string {
	return strings.ReplaceAll(s, "\r\n", "\n")
}

// findBestMatch finds the substring most similar to target.
func findBestMatch(text, target string) (string, float64) {
	lines := strings.Split(text, "\n")
	targetLines := strings.Split(target, "\n")

	if len(targetLines) == 1 {
		// Single line - search for similar line
		bestMatch := ""
		bestSimilarity := 0.0

		for _, line := range lines {
			sim := similarity(line, target)
			if sim > bestSimilarity {
				bestSimilarity = sim
				bestMatch = line
			}
		}
		return bestMatch, bestSimilarity
	}

	// Multi-line - search for similar block
	targetLen := len(targetLines)
	bestMatch := ""
	bestSimilarity := 0.0

	for i := 0; i <= len(lines)-targetLen; i++ {
		block := strings.Join(lines[i:i+targetLen], "\n")
		sim := similarity(block, target)
		if sim > bestSimilarity {
			bestSimilarity = sim
			bestMatch = block
		}
	}

	return bestMatch, bestSimilarity
}

// similarity calculates normalized Levenshtein similarity using the agnivade/levenshtein package.
// This provides better performance and handles edge cases more robustly than a custom implementation.
func similarity(a, b string) float64 {
	// Handle empty strings
	if len(a) == 0 && len(b) == 0 {
		return 1.0
	}
	if len(a) == 0 || len(b) == 0 {
		return 0.0
	}

	// For very long strings, use a rough approximation to avoid performance issues
	// The levenshtein package handles large strings well, but we still cap for extreme cases
	if len(a) > 10000 || len(b) > 10000 {
		// Simple length-based approximation for extremely long strings
		maxLen := max(len(a), len(b))
		minLen := min(len(a), len(b))
		return float64(minLen) / float64(maxLen)
	}

	dist := levenshtein.ComputeDistance(a, b)
	maxLen := max(len(a), len(b))
	return 1.0 - float64(dist)/float64(maxLen)
}

func (t *EditTool) EinoTool() einotool.InvokableTool {
	return &einoToolWrapper{tool: t}
}
